Marlin/src/module/stepper.cpp

```diff
@@ -1845,18 +1853,31 @@ void Stepper::pulse_phase_isr() {
       DELTA_ERROR = de; \
     }while(0)
 
+    // Differential Extruder moves E along with X
+    #if ENABLED(DIFFERENTIAL_EXTRUDER)
+      #define DIFFERENTIAL_E_PULSE(AXIS, ONSTEP) \
+        if (!is_print_job && _AXIS(AXIS) == X_AXIS) { \
+          E_APPLY_DIR(count_direction[X_AXIS] > 0, false); /* Match E direction to X */ \
```

Comment:
It is possible this DIR change will not be registered before the STEP is applied on the following line. Any direction change should include a delay so the driver can register the change before the next step signal. The stepper ISR currently aims to set all stepper DIR signals first, then delay, and then do all the steps. This way we only need one delay for all stepper DIRs. Any added delays will cause a reduction in the top stepping rate. But, we do need to add some kind of delay right here.

Reply:
Thanks. I'll take a look at the code you changed in stepper.cpp and test it on my printer.

Reply:
Homing and other commands issued via console/screen/click-wheel work ok.
Prints also come out ok, but that's up to planner.cpp

Reply:
There is an issue for which I do have a solution, but it could be more elegant. Here it is:
At the beginning of a gcode file, there is a G28 command ("home all axes"). When this command is actioned, I think the job timer is on, but for some reason planner.cpp does not get involved in calculating the pulses for steppers. G28 appears to go directly to stepper.cpp, which then processes it with no regard to Differential Extruder. The end result is that X and E steppers get out of sync and lots of skipping happens on the loop belt.
If I put an M76 command before G28, and then an M77 command after it, everything works fine.